File: PixelAdventure/PixelAdventure/GameViewController.m
//
//  GameViewController.m
//  PixelAdventure
//
//  Created by Jo Albright on 6/8/15.
//  Copyright (c) 2015 Jo Albright. All rights reserved.
//

#import "GameViewController.h"
#import "GameScene.h"

@implementation SKScene (Unarchive)

+ (instancetype)unarchiveFromFile:(NSString *)file {
    /* Retrieve scene file path from the application bundle */
    NSString *nodePath = [[NSBundle mainBundle] pathForResource:file ofType:@"sks"];
    /* Unarchive the file to an SKScene object */
    NSData *data = [NSData dataWithContentsOfFile:nodePath
                                          options:NSDataReadingMappedIfSafe
                                            error:nil];
    NSKeyedUnarchiver *arch = [[NSKeyedUnarchiver alloc] initForReadingWithData:data];
    [arch setClass:self forClassName:@"SKScene"];
    SKScene *scene = [arch decodeObjectForKey:NSKeyedArchiveRootObjectKey];
    [arch finishDecoding];
    
    return scene;
}

@end

@implementation GameViewController
{
    GameScene * scene;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    // Configure the view.
    SKView * skView = (SKView *)self.view;
    skView.showsFPS = YES;
    skView.showsNodeCount = YES;
    /* Sprite Kit applies additional optimizations to improve rendering performance */
    skView.ignoresSiblingOrder = YES;
    
    // Create and configure the scene.
    scene = [GameScene unarchiveFromFile:@"GameScene"];
    scene.scaleMode = SKSceneScaleModeResizeFill;
    
    // Present the scene.
    [skView presentScene:scene];
}

- (BOOL)shouldAutorotate
{
    return YES;
}

- (NSUInteger)supportedInterfaceOrientations
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        return UIInterfaceOrientationMaskAllButUpsideDown;
    } else {
        return UIInterfaceOrientationMaskAll;
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Release any cached data, images, etc that aren't in use.
}

- (IBAction)pressedDown:(UIButton *)sender {
    
    CGFloat direction = ([sender.titleLabel.text isEqualToString:@"A"]) ? -50 : 50;
    
    [scene changePixelDirection:direction];
    
}

- (IBAction)releasedButton:(UIButton *)sender {
    
    [scene changePixelDirection:0];
    
}


- (IBAction)moveButtonPressed:(UIButton *)sender {
    
    NSArray * letters = @[@"W",@"A",@"S",@"D"];
    
    int index = (int)[letters indexOfObject:sender.titleLabel.text];
    
    NSArray * directions = @[
                             [NSValue valueWithCGVector:CGVectorMake(0.0, 30.0)],
                             [NSValue valueWithCGVector:CGVectorMake(-10.0, 0.0)],
                             [NSValue valueWithCGVector:CGVectorMake(0.0, -10.0)],
                             [NSValue valueWithCGVector:CGVectorMake(10.0, 0.0)]
                             ];
    
    [scene movePixelInDirection:[directions[index] CGVectorValue]];
    
}

- (IBAction)fireButtonPressed:(UIButton *)sender {
    
    if ([sender.titleLabel.text isEqualToString:@"1"]) {
        
        [scene normalAttack];
        
    }
    
    if ([sender.titleLabel.text isEqualToString:@"2"]) {
        
        [scene specialAttack];
        
    }
    
}


- (BOOL)prefersStatusBarHidden {
    return YES;
}

@end

File: StuffedAnimalMortalKombat/GameViewController.swift
//
//  GameViewController.swift
//  StuffedAnimalMortalKombat
//
//  Created by Jo Albright on 6/9/15.
//  Copyright (c) 2015 Jo Albright. All rights reserved.
//

import UIKit
import SpriteKit
import MultipeerConnectivity

extension SKNode {
    class func unarchiveFromFile(file : String) -> SKNode? {
        if let path = NSBundle.mainBundle().pathForResource(file, ofType: "sks") {
            var sceneData = NSData(contentsOfFile: path, options: .DataReadingMappedIfSafe, error: nil)!
            var archiver = NSKeyedUnarchiver(forReadingWithData: sceneData)
            
            archiver.setClass(self.classForKeyedUnarchiver(), forClassName: "SKScene")
            let scene = archiver.decodeObjectForKey(NSKeyedArchiveRootObjectKey) as! GameScene
            archiver.finishDecoding()
            return scene
        } else {
            return nil
        }
    }
}

class GameViewController: UIViewController {
    
    var scene: GameScene?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        Connector.sharedConnector().gameBoard = self
        Connector.sharedConnector().startBrowsing()

        scene = GameScene.unarchiveFromFile("GameScene") as? GameScene

        if let scene = scene {
            // Configure the view.
            let skView = self.view as! SKView
            skView.showsFPS = true
            skView.showsNodeCount = true
            
            /* Sprite Kit applies additional optimizations to improve rendering performance */
            skView.ignoresSiblingOrder = true
            
            /* Set the scale mode to scale to fit the window */
            scene.scaleMode = .AspectFill
            
            skView.presentScene(scene)
        }
    }
    
    var playerViews: [PlayerStats] = []
    
    func playerJoined(peerID: MCPeerID) {
        
        println("player joined " + peerID.displayName)
        
        if let playerVC = storyboard?.instantiateViewControllerWithIdentifier("playerStats") as? UIViewController {
            
            let playerView = playerVC.view as! PlayerStats
            
            playerView.statsLives.lifeCount = 3
            playerView.playerName.text = peerID.displayName
            
            playerViews.append(playerView)
            
            scene?.playerJoined(peerID.displayName)
            
        }
        
        layoutPlayerStats()
        
    }
    
    func playerLeft(peerID: MCPeerID) {
        
        var foundPlayerViewIndex: Int?
        
        for (p,playerView) in enumerate(playerViews) {
            
            if let name = playerView.playerName.text, peerName = peerID.displayName where name == peerName {
                
                foundPlayerViewIndex = p
                scene?.playerLeft(name)
                
            }
            
        }
        
        // remove all playerViews
        for playerView in playerViews { playerView.removeFromSuperview() }
        
        if let foundPlayerViewIndex = foundPlayerViewIndex {
            
            playerViews.removeAtIndex(foundPlayerViewIndex)
            
        }
        
        
        layoutPlayerStats()
        
    }

    func layoutPlayerStats() {
        
        let padding: CGFloat = 20
        let maxPlayers: CGFloat = 6
        
        let playerStatsWidth = (view.frame.width - (padding * (maxPlayers + 1))) / maxPlayers
        
        for (p,playerView) in enumerate(playerViews) {
            
            view.addSubview(playerView)
            
            let x = (playerStatsWidth + padding) * CGFloat(p) + padding
            
            playerView.frame = CGRectMake(x, padding, playerStatsWidth, 300)
            
            println(playerView)
            println(playerView.healthBar.points)
            println(playerView.statsLives.lifeCount)
            
            view.layoutSubviews()
            
        }
        
    }
    
    
    // player movement functions
    
    // player attack functions

    override func shouldAutorotate() -> Bool {
        return true
    }

    override func supportedInterfaceOrientations() -> Int {
        if UIDevice.currentDevice().userInterfaceIdiom == .Phone {
            return Int(UIInterfaceOrientationMask.AllButUpsideDown.rawValue)
        } else {
            return Int(UIInterfaceOrientationMask.All.rawValue)
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Release any cached data, images, etc that aren't in use.
    }

    override func prefersStatusBarHidden() -> Bool {
        return true
    }
}
